### DISCUSSION TITLE
Do I have to declare all dependencies of an eslint-plugin as dependencies?

### DISCUSSION THREAD

**Discussion by gjgarryuan** [eslint/eslint#15681] [Q&A]
I am authoring a eslint plugin for my team which serves as the central place for all eslint dependencies and configurations in the wish that the consumer packages do not need to specify any of them other than simply "extends" of the configuration the plugin provides.

The plugin has the following package.json:

```
{
    "name": "@scope/eslint-plugin-team",
    ...
    "dependencies": {
        "@typescript-eslint/eslint-plugin": "5.11.0",
        "@typescript-eslint/parser": "5.11.0",
        "eslint": "8.9.0",
        "eslint-config-prettier": "8.3.0",
        "eslint-plugin-import": "2.25.4",
        "eslint-plugin-json-files": "1.3.0",
        "eslint-plugin-prettier": "4.0.0",
        "eslint-plugin-simple-import-sort": "7.0.0",
        "prettier": "2.5.1"
    },
    "peerDependencies": {
        "eslint": "^8.9.x",
        "prettier": "^2.5.x"
    }
}
```

I am not pasting the configs and rules code here as they are working, as it has been used by other packages in the same monorepo where all dependencies are hoisted.

When I tried to use the eslint plugin in an npm package of a different git repo "PackageA", I encountered dependency issues. 

Debug logs:

```
eslintrc:config-array-factory Loading package.json config file: /path/to/repos/PackageA/package.json +1ms
eslintrc:config-array-factory Loading JSON config file: /path/to/repos/PackageA/package.json +0ms
eslintrc:config-array-factory Config file found: /path/to/repos/PackageA/package.json +0ms
eslintrc:config-array-factory Loading {extends:"plugin:@scope/team/typescript"} relative to /path/to/repos/PackageA/package.json +0ms
eslintrc:config-array-factory Loading plugin "@scope/team" from /path/to/repos/PackageA/package.json +1ms
eslintrc:config-array-factory Loaded: @scope/eslint-plugin-team@1.0.0 (/path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/dist/lib/index.js) +0ms
eslintrc:config-array-factory Plugin /path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/dist/lib/index.js loaded in: 0ms +1ms
eslintrc:config-array-factory Loading {extends:"eslint:recommended"} relative to /path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/dist/lib/index.js +1ms
eslintrc:config-array-factory Loading JS config file: /path/to/repos/PackageA/node_modules/eslint/conf/eslint-recommended.js +1ms
eslintrc:config-array-factory Loading {extends:"plugin:@typescript-eslint/recommended"} relative to /path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/dist/lib/index.js +0ms
eslintrc:config-array-factory Loading plugin "@typescript-eslint" from /path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/dist/lib/index.js +1ms
eslintrc:config-array-factory Failed to load plugin '@typescript-eslint' declared in 'package.json » plugin:@scope/team/typescript#overrides[0]'. +0ms
```

As suggested by these debug logs particularly

```
eslintrc:config-array-factory Loading {extends:"plugin:@typescript-eslint/recommended"} relative to /path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/dist/lib/index.js +0ms
eslintrc:config-array-factory Loading plugin "@typescript-eslint" from /path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/dist/lib/index.js +1ms
```

eslint is trying to load "@typescript-eslint/eslint-plugin" from within /path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/dist/lib/, which is correct.

I verified that the plugin is visible to that folder by:

```
➜  ~ cd /path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/dist/lib
➜  lib git:(master) ✗ node
Welcome to Node.js v14.19.0.
Type ".help" for more information.
> require.resolve("@typescript-eslint/eslint-plugin")
'/path/to/repos/PackageA/node_modules/@scope/eslint-plugin-team/node_modules/@typescript-eslint/eslint-plugin/dist/index.js'
```

However, the error messages:

```
Oops! Something went wrong! :(

ESLint: 8.9.0

ESLint couldn't find the plugin "@typescript-eslint/eslint-plugin".

(The package "@typescript-eslint/eslint-plugin" was not found when loaded as a Node module from the directory "/path/to/repos/PackageA".)

It's likely that the plugin isn't installed correctly. Try reinstalling by running the following:

    npm install @typescript-eslint/eslint-plugin@latest --save-dev

The plugin "@typescript-eslint/eslint-plugin" was referenced from the config file in "package.json » plugin:@scope/team/typescript#overrides[0]".

If you still can't figure out the problem, please stop by https://eslint.org/chat/help to chat with the team.
```

indicates that it was trying to load "@typescript-eslint/eslint-plugin" from "/path/to/repos/PackageA", which definitely wouldn't work because "PackageA" does not declare "@typescript-eslint/eslint-plugin" as one of the devDependencies.

So my questions:
* debug loggings and error loggings have discrepancies, is it a bug?
* does PackageA have to declare all the dependencies used by @scope/eslint-plugin-team to work? (if so, this defeats the purpose of abstracting the eslint relevant dependencies and configurations to its own package)

**Comment by chiefmikey** [comment]
I believe you are running into this issue because all of your plugin dependencies are fixed versions so npm is nesting them in a node_modules within your package directory in the root node_modules. This is a problem because eslint resolves dependencies from the directory where the end user runs it and not from where the extended config that requires them is located. This behavior discourages global installations, something the eslint team seems to strongly oppose. 

The CLI has an option `--resolve-plugins-relative-to` and the Node API has `options.resolvePluginsRelativeTo` but those only set the path to where plugins are located and do not include custom parsers or shareable configs.

One solution would be to use compatible versions of the dependencies such as `"@typescript-eslint/eslint-plugin": "^5.11.0"` instead of `"@typescript-eslint/eslint-plugin": "5.11.0"`. Then they will be included in the root node_modules when installing your plugin and resolved correctly by eslint. 

If fixed versions are required then your best option might be to use the file path in your config when extending dependencies. [From the eslint docs](https://eslint.org/docs/user-guide/configuring/configuration-files#using-a-configuration-file): "The extends property value can be an absolute or relative path to a base configuration file." So in your case that could look like:

```js
module.exports = {
  extends: [
    path.resolve(
      path.dirname(
        require.resolve('../../node_modules/@typescript-eslint/eslint-plugin'),
      ),
      'configs/recommended.js',
    ),
    // ...
  ],
};
```

**Reply by gjgarryuan** [reply]
I will mark it as an answer once I verify your suggestion.

And thank you for the explanation it is really helpful, much appreciated!

**Reply by chiefmikey** [reply]
No problem, let me know if you run into any issues with it! Also just a quick suggestion, I personally have run into versioning issues when I included eslint as a dependency itself in my config package due to the other dependencies relying on differing versions of it, which resulted in errors because of the nesting issues again. If you're not getting any errors complaining about eslint's version you might be fine, but if it comes up in the future my solution was to just install eslint locally in each project along side my config package. Not as ideal as only having to install your plugin but it helps keep eslint from hitting errors. I do the same with prettier and stylelint, in general it seems like IDE extensions that rely on packages run better when they are installed locally in each project root.

**Reply by chiefmikey** [reply]
> I will mark it as an answer once I verify your suggestion.
> 
> 
> 
> And thank you for the explanation it is really helpful, much appreciated!

@garrychnca Just looking back at old discussions and wanted to check if you came across any other solutions to this? Thanks!

**Comment by liby** [comment]
@chiefmikey I have a similar issue.
<details>
<summary>Output</summary>

```
❯ yarn eslint
yarn run v1.22.15
$ eslint 'src/**/*.ts*'

Oops! Something went wrong! :(

ESLint: 7.29.0

ESLint couldn't find the plugin "eslint-plugin-prettier".

(The package "eslint-plugin-prettier" was not found when loaded as a Node module from the directory "/Users/user/Code/Company/project".)

It's likely that the plugin isn't installed correctly. Try reinstalling by running the following:

    npm install eslint-plugin-prettier@latest --save-dev

The plugin "eslint-plugin-prettier" was referenced from the config file in ".eslintrc#overrides[0] » @team-config/eslint-config-typescript/react".

If you still can't figure out the problem, please stop by https://eslint.org/chat/help to chat with the team.

error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
</details>

> One solution would be to use compatible versions of the dependencies such as "@typescript-eslint/eslint-plugin": "^5.11.0" instead of "@typescript-eslint/eslint-plugin": "5.11.0". Then they will be included in the root node_modules when installing your plugin and resolved correctly by eslint.

I am using the compatible version of `eslint-plugin-prettier`.

<details>
<summary>@team-config/eslint-config-typescript/package.json</summary>

```
{
  "name": "@team-config/eslint-config-typescript",
  "version": "1.0.18",
  "description": "ESLint Config for Team",
  "main": "index.js",
  "devDependencies": {
    "typescript": "^4.0.5"
  },
  "dependencies": {
    "@typescript-eslint/eslint-plugin": "^4.7.0",
    "@typescript-eslint/parser": "^4.7.0",
    "eslint": "7.29.0",
    "eslint-config-airbnb": "^18.2.1",
    "eslint-config-airbnb-base": "^14.2.1",
    "eslint-config-prettier": "^6.15.0",
    "eslint-import-resolver-typescript": "^2.3.0",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-jsx-a11y": "^6.4.1",
    "eslint-plugin-prettier": "^4.0.0",
    "eslint-plugin-react": "^7.24.0",
    "eslint-plugin-react-hooks": "^4.2.0",
    "prettier": "^2.4.1"
  }
}

```
</details>

After investigation, I found that this error has something to do with `prettier`.
<details>
<summary>project/package.json</summary>

```
{
  "name": "project",
  "version": "1.0.0",
  "scripts": {"eslint": "eslint 'src/**/*.ts*'"},
  "devDependencies": {
    ... // other dependencies
    "@team-config/eslint-config-typescript": "^1.0.18",
    "prettier-stylelint": "^0.4.2",
    "prettier": "^2.6.0"
  },
  "resolutions": {
    "**/prettier": "^2.5.1"
  },
  "engines": {
    "yarn": ">= 1.0.0"
  }
}
```
</details>

<details>
<summary>Output: yarn why prettier</summary>

```
❯ yarn why prettier
yarn why v1.22.15
[1/4] 🤔  Why do we have the module "prettier"...?
[2/4] 🚚  Initialising dependency graph...
[3/4] 🔍  Finding dependency...
[4/4] 🚡  Calculating file sizes...
=> Found "prettier@2.6.0"
info Has been hoisted to "prettier"
info Reasons this module exists
   - Specified in "devDependencies"
   - Hoisted from "@team-config#mobx-rest-axios-adapter#prettier"
   - Hoisted from "@team-config#eslint-config-typescript#prettier"
info Disk size without dependencies: "16.99MB"
info Disk size with unique dependencies: "16.99MB"
info Disk size with transitive dependencies: "16.99MB"
info Number of shared dependencies: 0
=> Found "@storybook/addon-storysource#prettier@2.3.0"
info This module exists because "@storybook#addon-storysource" depends on it.
info Disk size without dependencies: "19.01MB"
info Disk size with unique dependencies: "19.01MB"
info Disk size with transitive dependencies: "19.01MB"
info Number of shared dependencies: 0
=> Found "prettier-stylelint#prettier@1.19.1"
info This module exists because "prettier-stylelint" depends on it.
info Disk size without dependencies: "9.16MB"
info Disk size with unique dependencies: "9.16MB"
info Disk size with transitive dependencies: "9.16MB"
info Number of shared dependencies: 0
=> Found "@storybook/addon-docs#prettier@2.3.0"
info This module exists because "@storybook#addon-essentials#@storybook#addon-docs" depends on it.
info Disk size without dependencies: "19.01MB"
info Disk size with unique dependencies: "19.01MB"
info Disk size with transitive dependencies: "19.01MB"
info Number of shared dependencies: 0
=> Found "@storybook/source-loader#prettier@2.3.0"
info This module exists because "@storybook#addon-storysource#@storybook#source-loader" depends on it.
info Disk size without dependencies: "19.01MB"
info Disk size with unique dependencies: "19.01MB"
info Disk size with transitive dependencies: "19.01MB"
info Number of shared dependencies: 0
=> Found "@storybook/csf-tools#prettier@2.3.0"
info This module exists because "@storybook#addon-essentials#@storybook#addon-docs#@storybook#csf-tools" depends on it.
info Disk size without dependencies: "19.01MB"
info Disk size with unique dependencies: "19.01MB"
info Disk size with transitive dependencies: "19.01MB"
info Number of shared dependencies: 0
✨  Done in 0.78s.
```
</details>

<details>
<summary>The directory structure of the node_modules folder at this point is</summary>

```
.
├── ...
|   └── ...
├── team-config
|   └── ...
│   └── eslint-config-typescript
│       ├── base.js
│       ├── index.js
│       ├── javascript-node.js
│       ├── node.js
│       ├── node_modules
│       │   ├── .bin
│       │   ├── eslint-plugin-prettier
│       │   └── prettier // "version": "2.5.1",
│       ├── package.json
│       └── react.js
├── eslint
│   └── ...
├── eslint-config-airbnb
│   └── ...
├── eslint-config-airbnb-base
│   └── ...
├── eslint-config-prettier
│   └── ...
├── eslint-import-resolver-node
│   └── ...
├── eslint-import-resolver-typescript
│   └── ...
├── eslint-module-utils
│   └── ...
├── eslint-plugin-cypress
│   └── ...
├── eslint-plugin-import
│   └── ...
├── eslint-plugin-jsx-a11y
│   └── ...
├── eslint-plugin-react
│   └── ...
├── prettier // "version": "2.6.0",
│   └── ...
| ...
└── ...
    └── ...
```
</details>

When I remove `"**/prettier": "^2.5.1",` from `resolutions`, the *node_modules* folder remains unchanged.
<details>
<summary>yarn.lock file has changed</summary>

<img width="1848" alt="image" src="https://user-images.githubusercontent.com/38807139/159221731-901e3336-ad4d-46a4-9284-d1a7896c1616.png">
</details>

There is still an error when running the `yarn eslint` command.

After that, I tried to run the `yarn-deduplicate yarn.lock` command.
<details>
<summary>File yarn.lock has changed again</summary>

<img width="1809" alt="image" src="https://user-images.githubusercontent.com/38807139/159222379-4118dc44-e048-4157-beea-3e6fd0a105a9.png">
</details>

<details>
<summary>The current node_modules folder structure is as follows</summary>

```
.
├── ...
|   └── ...
├── team-config
|   └── ...
│   └── eslint-config-typescript
│       ├── base.js
│       ├── index.js
│       ├── javascript-node.js
│       ├── node.js
│       ├── node_modules
│       │   └── .bin
│       ├── package.json
│       └── react.js
├── eslint
│   └── ...
├── eslint-config-airbnb
│   └── ...
├── eslint-config-airbnb-base
│   └── ...
├── eslint-config-prettier
│   └── ...
├── eslint-import-resolver-node
│   └── ...
├── eslint-import-resolver-typescript
│   └── ...
├── eslint-module-utils
│   └── ...
├── eslint-plugin-cypress
│   └── ...
├── eslint-plugin-import
│   └── ...
├── eslint-plugin-jsx-a11y
│   └── ...
├── eslint-plugin-prettier
│   └── ...
├── eslint-plugin-react
│   └── ...
├── prettier // "version": "2.6.0",
│   └── ...
| ...
└── ...
    └── ...
```
</details>

Running the `yarn eslint` command did not see any errors. But I still don't understand why.

**Reply by chiefmikey** [reply]
@liby Sorry for my delayed reply. I don't have a great answer as to why this solution works but I do believe it still has to do with the nested dependencies. I have personally had issues with including main packages like eslint/prettier/stylelint as dependencies in my config packages. By including prettier as a dependency in your @team-config/eslint-config-typescript/package.json it creates a nested package as seen in your first node_modules tree when the error was occurring. Since yarn-deduplicate removes previous versions of dependencies when a more current one exists that meets all version requirements it is removing the nested prettier package in favor of the newer version in the root node_modules which then gave no error. I would try removing prettier from the dependencies of @team-config/eslint-config-typescript/package.json and keeping the local install in project/package.json. You may need to delete your node_modules directory along with the yarn.lock file and do a fresh install once updated.

**Reply by chiefmikey** [reply]
Thinking about it more, the error is being thrown when eslint-plugin-prettier is being nested because eslint is being run from the root directory of project so it's only looking in the root node_modules for dependencies. I'm not 100% certain why but eslint-plugin-prettier must be linking to the nested prettier package included with it in @team-config/eslint-config-typescript resulting in the plugin being nested along side it. After running yarn-deduplicate and removing the nested prettier directory the eslint-plugin-prettier package is relocated to the root node_modules where eslint is able to resolve it.

**Reply by liby** [reply]
I'm sorry, I've been on vacation lately. I just saw your reply.

> I would try removing prettier from the dependencies of "@team-config/eslint-config-typescript/package.json" and keeping the local install in "project/package.json".

It seems that this is the recommended practice for such projects. The main reason for not doing this at the time was that each project that used the "eslint/prettier/eslint-plugin-xxx" dependency would have to be installed repeatedly, which would be a pain in the ass when upgrading the dependency.
Now it seems that a uniform installation in a configuration project brings more incomprehensible troubles instead.

> the error is being thrown when "eslint-plugin-prettier" is being nested because `eslint` is being run from the root directory of project so it's only looking in the root *node_modules* for dependencies. I'm not 100% certain why but "eslint-plugin-prettier" must be linking to the nested prettier package included with it in "@team-config/eslint-config-typescript" resulting in the plugin being nested along side it

Yes, this is exactly what I have observed, but I just can't get a good explanation, and I don't know if the subsequent migration to PNPM will solve this issue.
